doc/reference/trd-userspace-readable-allow-syscalls.md

```diff
@@ -0,0 +1,183 @@
+The buffer number specifies which buffer this is. A driver may
+support multiple allowed buffers.
+
+The return variants for Read-Write Allow system calls are `Failure
+with 2 u32` and `Success with 2 u32`.  In both cases, `Argument 0`
+contains an address and `Argument 1` contains a length. When a driver
+implementing the Read-Write Allow system call returns a failure
+result, it MUST return the same address and length as those that were passed
+in the call. When a driver implementing the Read-Write Allow system call
+returns a success result, the returned address and length MUST be those
+that were passed in the previous call, unless this is the first call.
+On the first successful invocation of a particular Read-Write Allow system
+call, an driver implementation MUST return address 0 and size 0.
+
+The syscall class ID is shown below:
+
+| Syscall Class    | Syscall Class Number |
+|------------------|----------------------|
+| Read-Write Allow |           7          |
+
+The standard access model for userspace readable allowed buffers is that userspace can read
+from a buffer while the kernel can read or write. Syncronisation
+methods are required to ensure data consistency but are implementation specifc.
+
+Simultaneous accesses to a buffer from both userspace and the kernel can
+cause issues if not implemented properly. For example a userspace app could
+read partially written data.
+This would result in obscure timing bugs that are hard to detect. Due to this
+each capsule using the userspace readable Allow mechanism MUST document, in a Draft or
+Final Documentary TRD, how consistency of the data exchanged through this
+userspace readable memory region is achieved.
+
+Finally, because a process conceptually relinquishes write access to a buffer
+when it makes a userspace readable allow call with it, a userspace API MUST NOT
+assume or rely on a process writing an allowed buffer. If userspace needs to
+write to a buffer held by the kernel, it MUST first regain access to it by
+calling the corresponding Userspace Readable Allow. A userspace API MAY
+allow a process to read an allowed buffer, but if it does, it must document
+a consistency mechanism.
+
+One example approach to ensure that userspace reads of a data object are
+consistent is to use a monotonic counter. Every time the kernel writes
+the data object, it increments the counter. If userspace reads the counter,
+reads the data object, then reads the counter again to check that it has not
+changed, it can check that the object was not modified mid-read. If the counter
+changes, it restarts the read of the data object. This approach is simple, but
+does make reading the data object take variable time and is theoretically
+vulnerable to starvation.
+
+An example of reading a monotomic counter from userspace would look like this:
+
+```c
+  volatile uint32_t high, low;
+
+  do {
+    // Set the high bytes the value in memory
+    high = ptr[3];
+    // Read the low bytes
```

Comment:
Please add the actual data read in here; this example is just for a counter to be extended to general data as described in the above paragraph.